PDSim/core/core.py

```diff
@@ -1131,39 +1131,82 @@ def OBJECTIVE_CYCLE(self, Td_Tlumps0, X, epsilon = 0.003, cycle_integrator = 'RK
-                Deltat = (2*np.pi)/self.omega # [s]
-            
-                #  Update the lump temperatures
-                self.Tlumps = [_T + _EB*Deltat/thermal_capacitance]
-            
+            if len(self.Tlumps) > 1:
+                print("Running multi-lump analysis")
+
+                if self.OEB_solver == 'MDNR':
+                    # Use Multi Dim. Newton Raphson step for multi-lump temperatures
+                    from numpy.linalg import inv
+                    w = 1.0                
+                    dx = 0.5
+                    ytol=1e-3
+                    x = np.array(self.Tlumps,dtype=np.float)
+                    J = np.zeros((len(x),len(x)))
+                    error = 999
+                    # If a float is passed in for dx, convert to a numpy-like list the same shape
+                    # as x
+                    if isinstance(dx,int) or isinstance(dx,float):
+                        dx=dx*np.ones_like(x)
+                                
+                    r0 = np.array(self.lumps_resid)*1000 
+
+                    # Build the Jacobian matrix by columns
+                    for jj in range(len(self.Tlumps)):
+                        delta = np.zeros_like(x)
+                        delta[jj] = dx[jj]
+                        self.Tlumps = self.Tlumps + delta
+                        ri = self.callbacks.lumps_energy_balance_callback()
+                        #print('ri:',ri)
+                        ri = np.array(ri)
+                        J[:,jj] = (ri-r0)/delta[jj]
+
+                    v = np.dot(-inv(J),r0)
+
+                    # Calculate new Tlumps
+                    Tnew = x + w*v
+                    self.Tlumps = Tnew
+
+                elif self.OEB_solver == 'Broyden':
+                    # Use Broyden Method  
+                    raise('Broyden not implemented yet')
+
             else:
+                # Use Relaxed Secant Method for single lump temperature
+                print("Running single-lump analysis")
                 
-                #  Get the values from the history
-                Tn1, EBn1 = self.solvers.lump_eb_history[-1]
-                Tn2, EBn2 = self.solvers.lump_eb_history[-2]
+                if len(self.solvers.lump_eb_history) == 1:
+                    
+                    T, EB = self.solvers.lump_eb_history[-1]
+                    
+                    #  T and EB are one-element lists, get floats
+                    _T, _EB = T[0], EB[0]
                 
-                #  Convert to numpy arrays
-                Tn1, EBn1, Tn2, EBn2 = [np.array(l) for l in [Tn1, EBn1, Tn2, EBn2]]
+                    # Use the thermal mass to make the step
+                    # Here is the logic:
+                    # Instantaneous energy balance given by
+                    # dU/dt = m*c*(dT/dt) = sum(Qdot)
+                    # and if dt = one cycle period (seconds/rev) DeltaT = 2*pi/omega
```

Comment:
This was my mistake initially, but we should tidy up the text, and remove the DeltaT in the middle of the line

Reply:
Can you please change this one and then I'll merge?

Reply:
yes sure. I will fix all the comments. I spotted another tiny issue.